src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -0,0 +1,318 @@
+      skatStat += st(i).q
+      i += 1
+    }
+
+    val weightedGenotypes = new DenseMatrix[Double](n, m, xwArray)
+    val qtWeightedGenotypes = new DenseMatrix[Double](k, m, qtxwArray)
+    val GwGrammian = weightedGenotypes.t * weightedGenotypes
+    val QtGwGrammian = qtWeightedGenotypes.t * qtWeightedGenotypes
+
+    val SPG = new SkatModel(skatStat / (2 * sigmaSq))
+    SPG.computeLinearSkatStats(GwGrammian, QtGwGrammian)
+  }
+
+  def largeNResultOp[T <: Vector[Double]](st: Array[SkatTuple[T]], sigmaSq: Double): SkatStat = {
+    val m = st.length
+    val n = st(0).xw.size
+    val k = st(0).qtxw.size
+
+    val ZGrammianArray = new Array[Double](m * m)
+    val QtZGrammianArray = new Array[Double](m * m)
+
+    var i = 0
+    var j = 0
+
+    while (i < m) {
+      ZGrammianArray(i * m + i) = st(i).xw dot st(i).xw
+      j = 0
+      while (j < i) {
+        val ijdotprod = st(i).xw dot st(j).xw
+        ZGrammianArray(i * m + j) = ijdotprod
+        ZGrammianArray(j * m + i) = ijdotprod
+        j += 1
+      }
+      i += 1
+    }
+
+    i = 0
+    while (i < m) {
+      QtZGrammianArray(i * m + i) = st(i).qtxw dot st(i).qtxw
+      j = 0
+      while (j < i) {
+        val ijdotprod = st(i).qtxw dot st(j).qtxw
+        QtZGrammianArray(i * m + j) = ijdotprod
+        QtZGrammianArray(j * m + i) = ijdotprod
+        j += 1
+      }
+      i += 1
+    }
+
+    //compute the variance component score
+    var skatStat = 0.0
+    i = 0
+    while (i < m) {
+      skatStat += st(i).q
+      i += 1
+    }
+
+    val weightedGenotypesGrammian = new DenseMatrix[Double](m, m, ZGrammianArray)
+    val qtWeightedGenotypesGrammian = new DenseMatrix[Double](m, m, QtZGrammianArray)
+
+    val SPG = new SkatModel(skatStat / (2 * sigmaSq))
+    SPG.computeLinearSkatStats(weightedGenotypesGrammian, qtWeightedGenotypesGrammian)
+  }
+
+  def keyedRDDSkat[T <: Vector[Double]](vds: VariantDataset,
+    variantKeys: String,
+    singleKey: Boolean,
+    weightExpr: Option[String],
+    yExpr: String,
+    covExpr: Array[String],
+    getGenotypes: (Iterable[Genotype], Int) => T):
+  (RDD[(Any, Iterable[(T, Double)])], Type, DenseVector[Double], DenseMatrix[Double]) = {
+
+    val (y, cov, completeSamples) = RegressionUtils.getPhenoCovCompleteSamples(vds, yExpr, covExpr)
+
+    val n = y.size
+
+    val filteredVds = weightExpr match {
+      case None => vds.filterSamplesList(completeSamples.toSet)
+        .annotateVariantsExpr("va.AF = gs.callStats(g=> v).AF")
+        .annotateVariantsExpr("va.weight = let af = if (va.AF[0] <= va.AF[1]) va.AF[0] else va.AF[1] in dbeta(af,1.0,25.0)**2")
+      case _ => vds.filterSamplesList(completeSamples.toSet)
+    }
+
+    val (keysType, keysQuerier) = filteredVds.queryVA(variantKeys)
+    val (weightType, weightQuerier) = weightExpr match {
+      case None => filteredVds.queryVA("va.weight")
+      case Some(expr) => filteredVds.queryVA(expr)
+    }
+
+    val typedWeightQuerier = weightType match {
```

Comment:
returning null is leading to an implied type of Any which is causing an error later on. Is the null necessary?